rllib/evaluation/collectors/simple_list_collector.py

```diff
@@ -232,131 +233,51 @@ def build(self, view_requirements: ViewRequirementsDict) -> SampleBatch:
-                    data = [
-                        np.asarray(
-                            [
-                                d[
-                                    self.shift_before
-                                    + (i * view_req.batch_repeat_value)
-                                    + shift
-                                ]
-                                for i in range(count)
-                            ]
-                        )
-                        for d in np_data[data_col]
-                    ]
-                # Shift is exactly 0: Use trajectory as is.
-                elif shift == 0:
-                    data = [d[self.shift_before :] for d in np_data[data_col]]
-                # Shift is positive: We still need to 0-pad at the end.
-                elif shift > 0:
-                    data = [
-                        _to_float_np_array(
-                            np.concatenate(
-                                [
-                                    d[self.shift_before + shift :],
-                                    [
-                                        np.zeros(
-                                            shape=view_req.space.shape,
-                                            dtype=view_req.space.dtype,
-                                        )
-                                        for _ in range(shift)
-                                    ],
-                                ]
+
+                    # handle the case where the inds are out of bounds from the end
+                    element_at_t = []
+                    for index in inds:
+                        if index < len(d):
+                            element_at_t.append(d[index])
+                        else:
+                            element_at_t.append(
+                                np.zeros(
+                                    shape=view_req.space.shape,
+                                    dtype=view_req.space.dtype,
+                                )
                             )
-                        )
-                        for d in np_data[data_col]
-                    ]
-                # Shift is negative: Shift into the already existing and
-                # 0-padded "before" area of our buffers.
+                    element_at_t = np.stack(element_at_t)
+
+                    if element_at_t.shape[0] == 1:
+                        # squeeze to remove the T dimension if it is 1.
+                        element_at_t = element_at_t.squeeze(0)
+                    shifted_data.append(element_at_t)
+
+                # in some multi-agent cases shifted_data may be an empty list.
+                if shifted_data:
+                    shifted_data_np = np.stack(shifted_data, 0)
                 else:
-                    data = [
-                        d[self.shift_before + shift : shift] for d in np_data[data_col]
-                    ]
+                    shifted_data_np = np.array(shifted_data)
```

Comment:
don't get it man, if shifted_data is empty, why do we bother with np.array(shifted_data)?
if it's not empty, why do we stack a 0 at the end??
😵

Reply:
Read the code again :)). We are not stacking zero to the end. We are stacking a list of np.arrays at **index=0**. In case of an empty list, we just return an empty np.array which is apparently the expected external behavior.

Reply:
ok, I need to take another closer look at this.
but super nice. thanks a lot for the work man, what do I do without you :)

Reply:
yeah I totally understand this now. cool.